runtime/expr/values.go

```diff
@@ -177,61 +177,83 @@ func (r *recordSpreadExpr) invalidate(object map[string]column) {
 	r.cache = r.zctx.MustLookupTypeRecord(r.columns)
 }
 
-type ArrayExpr struct {
-	exprs []Evaluator
-	zctx  *zed.Context
+type VectorElem struct {
+	Value  Evaluator
+	Spread Evaluator
+}
 
-	builder zcode.Builder
-	elems   collectionBuilder
+type ArrayExpr struct {
+	builder    zcode.Builder
+	collection collectionBuilder
+	elems      []VectorElem
+	zctx       *zed.Context
 }
 
-func NewArrayExpr(zctx *zed.Context, exprs []Evaluator) *ArrayExpr {
+func NewArrayExpr(zctx *zed.Context, elems []VectorElem) *ArrayExpr {
 	return &ArrayExpr{
-		exprs: exprs,
+		elems: elems,
 		zctx:  zctx,
 	}
 }
 
 func (a *ArrayExpr) Eval(ectx Context, this *zed.Value) *zed.Value {
 	a.builder.Reset()
-	a.elems.reset()
-	for _, e := range a.exprs {
-		a.elems.append(e.Eval(ectx, this))
+	a.collection.reset()
+	for _, e := range a.elems {
+		if e.Value != nil {
+			a.collection.append(e.Value.Eval(ectx, this))
+			continue
+		}
+		val := e.Spread.Eval(ectx, this)
+		inner := zed.InnerType(val.Type)
+		if inner == nil {
+			return a.zctx.WrapError("spread syntax requires array/set", val)
```

Comment:
`recordSpeadExpr.Eval` doesn't create an error when asked to spread a non-record value; it just ignores that value. I'm not arguing in favor of that behavior over what's here but rather that what's here should be eventually be consistent with what's there.

My very mild preference would be to add new this new functionality with consistent behavior and then hash out different behavior in a separate PR, but I'm totally fine hashing it out in this PR and then updating `recordSpreadExpr.Eval` in another PR. Your call.

Reply:
Yeah it feels above my pay grade to make this call. I'll do as you suggest because I agree consistency should be the precedent when things are unclear like this.